### PR TITLE
feat(xds): unified-name for empty listener tags

### PR DESCRIPTION
**Pull request by lukidzi**
## Motivation

`MeshProxyPatch` selects generated Envoy listeners with `listenerTags`, matching key/value pairs Kuma writes into listener metadata under `io.kuma.tags`. As identity moved out of tags and into resource labels (`MeshService`, `InboundTagsDisabled`), those listeners started writing an empty `io.kuma.tags: {}`. The policy still validates and is accepted, but matches nothing — a silent regression on a user-facing API.

## Implementation information

Fills `io.kuma.tags` wherever it would otherwise be empty with a synthesized `kuma.io/unified-name` tag (MADR 105, sub-option B). The value is direction-specific:

- Outbound listeners get the destination `Mesh*Service` KRI, with the port as section name, via a new `DestinationService.OutboundListenerTags()` used by both the `meshtcproute` and `meshhttproute` generators. Legacy outbounds keep their real tags with `kuma.io/mesh` stripped, as before.
- Inbound listeners whose tags are empty get a stable self-reference, `self_inbound_dp_<sectionName>` — the same contextual name already used as the inbound's unified listener name (`naming.ContextualInboundName`) — via a new `generator.InboundListenerTags()` wired into both `inbound_proxy_generator` and the `meshtls` plugin. Inbounds that carry tags keep them unchanged.

The inbound value is deliberately not the `Dataplane` KRI: on Kubernetes the `Dataplane` name is the Pod name, so a KRI there carries the Pod hash and rotates on every rollout, which cannot be written ahead of time and would reintroduce the silent match-nothing failure. A `MeshProxyPatch` already selects the proxy through `spec.targetRef`, so the inbound match only has to disambiguate the port, which the section name does.

The rule keys on the tags being empty, not on any flag, so a hand-written Universal `Dataplane` with tagless inbounds is covered too. There is no `MeshProxyPatch` API change and the fill is unconditional. Because it can only add matches, nothing that matches today can stop matching.

Policies that matched `kuma.io/service: <name>` have to be rewritten against `kuma.io/unified-name`: the destination KRI on an outbound, `self_inbound_dp_<port>` on an inbound.

Tests: unit tests for both new helpers, an end-to-end golden for the emptied inbound path, plus the mechanical golden updates where outbound listeners now carry the synthesized KRI.

## Supporting documentation

- MADR 105: Source Envoy listener tags from resource labels
- Closes part of https://github.com/kumahq/kuma/issues/17381

### CODE REVIEW

File: pkg/xds/generator/inbound_proxy_generator.go [hunk 1]
```diff
@@ -190,6 +192,20 @@ func FilterChainBuilder(
...
 }
 
+// InboundListenerTags returns the inbound listener's io.kuma.tags. When the
+// inbound has no tags, it synthesizes a kuma.io/kri tag from the Dataplane KRI
+// (port as section name) so the listener stays selectable.
+func InboundListenerTags(dataplane *core_mesh.DataplaneResource, tags map[string]string, portName string) map[string]string {
+	if len(tags) > 0 {
+		return tags
+	}
+	id := kri.WithSectionName(kri.FromResourceMeta(dataplane.GetMeta(), core_mesh.DataplaneType), portName)
```
Comment (Automaat) [inline pkg/xds/generator/inbound_proxy_generator.go]: **blocking:** Two tagless inbounds on the same Dataplane without an explicit `name` synthesize the identical `kuma.io/kri` value — Mesh/Zone/Namespace/Name come only from the shared Dataplane meta and `SectionName` is empty for both — so a `MeshProxyPatch.listenerTags` selector meant for one inbound also matches the other's listener. Fall back to the port number as section name when `portName` is empty.

_Generated by Sybra harness_

File: pkg/plugins/policies/core/xds/meshroute/listeners.go [hunk 2]
```diff
@@ -84,6 +84,16 @@ func (ds *DestinationService) ConditionallyResolveKRIWithFallback(condition bool
 	return fallback
 }
 
+// OutboundListenerTags returns the outbound listener's io.kuma.tags: real
+// outbound tags without kuma.io/mesh for a legacy outbound, or a synthesized
+// kuma.io/kri tag from the destination KRI for a resource-based one.
+func (ds *DestinationService) OutboundListenerTags() envoy_tags.Tags {
```
Comment (Automaat) [inline pkg/plugins/policies/core/xds/meshroute/listeners.go]: **suggestion:** `OutboundListenerTags` (outbound, resource-based) and `InboundListenerTags` (inbound, Dataplane-based) implement the same "empty tags → synthesize KRI, else keep real tags" rule independently, with different signatures and return types (`envoy_tags.Tags` vs `map[string]string`). Worth extracting one shared "empty tags → KRI tag" helper both wrap, so the next call site doesn't reinvent this a third time.

_Generated by Sybra harness_

File: pkg/xds/generator/inbound_proxy_generator.go [hunk 3]
```diff
@@ -190,6 +192,20 @@ func FilterChainBuilder(
...
+	if id.IsEmpty() {
+		return tags
+	}
+	return map[string]string{mesh_proto.KRITag: id.String()}
```
Comment (Automaat) [inline pkg/xds/generator/inbound_proxy_generator.go]: **suggestion:** Neither `OutboundListenerTags`'s KRI branch nor `InboundListenerTags`'s KRI-fallback branch emits a log line or metric when taken. Since this is exactly the branch that silently changes what a `MeshProxyPatch` selector matches, an operator debugging "my patch stopped applying" has to diff Envoy config dumps by hand. A debug-level log keyed by dataplane/listener would make this diagnosable.

_Generated by Sybra harness_

File: pkg/plugins/policies/core/xds/meshroute/listeners.go [hunk 4]
```diff
@@ -84,6 +84,16 @@ func (ds *DestinationService) ConditionallyResolveKRIWithFallback(condition bool
 	return fallback
 }
 
+// OutboundListenerTags returns the outbound listener's io.kuma.tags: real
+// outbound tags without kuma.io/mesh for a legacy outbound, or a synthesized
+// kuma.io/kri tag from the destination KRI for a resource-based one.
+func (ds *DestinationService) OutboundListenerTags() map[string]string {
+	if id, ok := ds.Outbound.AssociatedServiceResource(); ok {
```
Comment (lukidzi) [inline pkg/plugins/policies/core/xds/meshroute/listeners.go:94]: Suggestion: OutboundListenerTags accesses ds.Outbound without a nil guard, unlike ConditionallyResolveKRIWithFallback above. CollectServices guarantees non-nil, so this won't panic in practice — but adding `if ds.Outbound == nil { return nil }` would make the invariant explicit and protect against future callers that skip CollectServices.

File: pkg/plugins/policies/core/xds/meshroute/listeners.go [hunk 5]
```diff
@@ -84,6 +84,16 @@ func (ds *DestinationService) ConditionallyResolveKRIWithFallback(condition bool
...
+		return map[string]string{mesh_proto.KRITag: id.String()}
+	}
+	return map[string]string(envoy_tags.Tags(ds.Outbound.TagsOrNil()).WithoutTags(mesh_proto.MeshTag))
```
Comment (lukidzi) [inline pkg/plugins/policies/core/xds/meshroute/listeners.go:97]: Suggestion: When a legacy outbound has nil or empty tags, this returns an empty map. InboundListenerTags synthesizes a KRI fallback when tags are empty. Is the asymmetry intentional? If MeshProxyPatch should still be able to select a tagless legacy outbound by KRI, this path may need the same fallback logic.

File: pkg/plugins/policies/core/xds/meshroute/listeners_test.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+		ds := meshroute.DestinationService{Outbound: &xds_types.Outbound{Resource: id}}
+
+		// when
+		tags := ds.OutboundListenerTags()
+
+		// then
+		Expect(tags).To(BeEquivalentTo(map[string]string{mesh_proto.KRITag: id.String()}))
+	})
+
+	It("returns the real outbound tags without kuma.io/mesh for a legacy outbound", func() {
```
Comment (lukidzi) [inline pkg/plugins/policies/core/xds/meshroute/listeners_test.go:33]: Suggestion: Tests cover real-resource outbounds and legacy outbounds with tags, but not the edge case of a legacy outbound with Tags: nil or Tags: map[string]string{}. Adding that case would document the current (empty-map) behaviour explicitly.

**Review by Automaat — CHANGES_REQUESTED** [review]
Solid root-cause fix for the empty io.kuma.tags regression — confirmed it cannot make any currently-matching MeshProxyPatch selector stop matching. One blocking bug: two tagless, unnamed inbounds on the same Dataplane synthesize the identical kuma.io/kri value, so a listenerTags selector for one silently also matches the other listener. A couple of non-blocking suggestions on consolidating the duplicate outbound/inbound tag-fill logic and adding observability for the tag-fill branch.

_Generated by Sybra harness_

File: pkg/xds/generator/inbound_proxy_generator.go (hunk 1 above)
Comment (lukidzi) [reply pkg/xds/generator/inbound_proxy_generator.go]: Fixed, though by dropping the KRI approach for inbounds rather than adding a port fallback. Inbound listeners now use `naming.MustContextualInboundName(dataplane, endpoint.InboundName)`, and `InboundName` comes from `Dataplane_Networking_Inbound.GetSectionName()`, which already falls back to the stringified port when `name` is empty. Two tagless inbounds on the same Dataplane therefore get `self_inbound_dp_8080` and `self_inbound_dp_9090` — distinct, so a `listenerTags` selector can no longer match both.

File: pkg/plugins/policies/core/xds/meshroute/listeners.go (hunk 2 above)
Comment (lukidzi) [reply pkg/plugins/policies/core/xds/meshroute/listeners.go]: No longer applies after the rework. The two functions do not implement the same rule anymore, so there is nothing shared to extract. `OutboundListenerTags` branches on real-resource vs legacy outbound (real resource gets the destination KRI, legacy keeps its own tags minus `kuma.io/mesh`), while `InboundListenerTags` branches on tag emptiness and falls back to the contextual name. Both also return `map[string]string` now, so the signature mismatch is gone too.

File: pkg/xds/generator/inbound_proxy_generator.go (hunk 3 above)
Comment (lukidzi) [reply pkg/xds/generator/inbound_proxy_generator.go]: Not adding logging here. The fallback is deterministic, not conditional on cluster state: the tag is `kuma.io/unified-name` set to the destination KRI on outbounds and to the contextual inbound name on inbounds, and it is visible directly in the listener metadata of an Envoy config dump. Unit tests and the updated golden files pin both branches, so a debug line would only restate what the config dump already shows.

**Review by Automaat — APPROVED** [review]
